src/main/java/org/eclipse/uprotocol/streamer/UStreamer.java

```diff
@@ -0,0 +1,183 @@
+     * 
+     * @param in input {@code Route} that the streamer listens to
+     * @param out output {@code Route} that the streamer forwards the message to
+     * @return {@code UStatus} with UCode.OK if the forwarding rule was added successfully,
+     *        {@code UStatus} with UCode.ALREADY_EXISTS if the forwarding rule already exists,
+     *       {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+     */
+    public UStatus addForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        // Cannot route to itself
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
+        }
+
+        // check if the rule already exists in the list
+        if (listeners.stream().anyMatch(p -> p.getInputRoute().equals(in) && p.getOutputRoute().equals(out))) {
+            return UStatus.newBuilder().setCode(UCode.ALREADY_EXISTS).build();
+        }
+
+        // Setup the forwarding rule for any messages who's destination (sink) is 
+        // out authority with the in uTransport
+        TransportListener listener = new TransportListener(in, out);
+        UUri out_uri = UUri.newBuilder().setAuthority(out.getAuthority()).build();
+        UStatus result = in.getTransport().registerListener(out_uri, listener);
+        if (result.getCode() != UCode.OK) {
+            return result;
+        }
+
+        listeners.add(listener);
+
+
+        // Published messages do not contain a sink so we cannot setup simple static forwarding rules based
+        // on the sink address like notifications, requests and responses. Registering a listener for all
+        // published events would mean that the streamer gets each and ever single published topic on the IN
+        // Route (not what we want). What we need to know is what published topics does the streamer really 
+        // need to care about (i.e register a listener for) so we don't get everything, this is only published
+        // topics that there is a subscriber in OUT UAuthority that has subscribed to said topic.
+        //
+        // The good thing is that the subscribers information (URI) is recorded in usubscription service when 
+        // the subscriber calls subscribe(), all we need to do is fetch ths list of subscriptions that match the 
+        // IN UAuthority when the subscriber is the OUT UAuthority and we have the magic list we need!
+        FetchSubscriptionsRequest request = FetchSubscriptionsRequest.newBuilder()
+            .setTopic(UUri.newBuilder().setAuthority(in.getAuthority()))
+            .setSubscriber(SubscriberInfo.newBuilder().setUri(UUri.newBuilder().setAuthority(out.getAuthority())))
+            .build();
+
+        // Get the list of subscriptions
+        FetchSubscriptionsResponse response = submgr.fetchSubscriptions(request);
```

Comment:
what happens if the subscription manager is slow - maybe under load?
what happens if the subscription manager fails? A bug or it is restarting or something?
Is this even a thing we need to handle in the car? Cloud applications must handle these cases and be a little more resilient.
Just saying that this is the happy path, subscription manager is a separate component and anything can happen.
The good thing is that you made it an interface and can now create a USubscription implementation that barfs and see how your application handles these use cases.